move transition to feedforward

ctzsnooze opened the pull request:
@etracer65 raised a concern that Rx noise from un-smoothed rcCommand might cause problems with feedforward transition.

Feedforward transition is a direct attenuator, by simple multiplication, of feedforward, derived from the absolute stick position.  The user sets a limit in percent from centre over which the attenuation should occur.

When sticks are centred, the multiplier is 0, rising linearly to 1 at the end of the transitional zone.

Since it is a simple multiplier, a step change in un-smoothed rcCommand of say 10% should cause a 10% change in feed forward.  And that should result in a step change of 10% in the pilot's feed forward value.  
However it is tough get such a big change from RC step to RC step with a modern radio, other than at 50hz or if there are dropouts.  For most normal radios, in normal operation, and certainly for any link above 100hz, very large changes from packet to packet require very rapid movement of the stick.

When the change is large, like that, so is the change in feed forward.  

Feedforward is calculated as steps directly from the un-smoothed setpoint values; hence it will immedialely change when every new RC value comes in.  Feedforward smoothing smooths these steps out, just as setpoint smoothing smooths out the steps in the setpoint signal.

So the worry was that the Transition effect may put sharp square steps into the otherwise smooth feedforward signal whenever the sticks move a lot.  The 'print-through' of transition noise would only happen if rcCommand consisted of steep, sharp steps.  And since #10723, these steps have not been smoothed.

The problem should be greatest when the transition zone is narrow, since the percentage change per RC step can be greater when the transition line is steeper .  However in most setups, a narrow transition range means that the effect would only apply in the centre, low expo region, where setpoint changes are small.  So those effects cancel out.

If the transition zone is wide, then there is relatively little change per step, so the potential for this problem is attenuated.

If we look hard at the log, and this is a problem, we should see the sharp step corners appearing superimposed on an otherwise smooth feedforward trace.  

I looked hard, and couldn't find any trace of this actually happening, not with any transition value or any feedforward value.

I looked really hard, and just couldn't see it, not with a standard FrSky link, and no matter how heavily I smoothed the feedforward trace.

However the possibility does exist that some rcCommand data may have extreme jitter.

And the transition maths is repeated every PID loop, even though actually the radio input signals only change infrequently.

We could calculate the transition amount within the feedforward code, when new RC data arrives, instead of every PID loop.  This way any step change in transition just multiplies the feedforward value before it is smoothed.  Once smoothed, there is no chance that the transition effect will result in any sharp corners in the feedforward trace.

This PR does just that, effectively eliminating the possibility of un-smoothed transfer of sharp rcCommand steps into feedforward.

However, it is a solution in search of a problem.  It solves nothing, to be honest.  There is just nothing to solve.

This graphic shows current master with un-smoothed rcCommand compared to the new code on the right.  Transition is 100, setpoint is un-smoothed to show us where the steps are.  There is no sign of any sharp stepping on the master code.  Both logs appear identical.  There are significant irregularities in feed forward due to my tremor, timer inaccuracy and other things, like in all feedforward traces.  These errors vastly outweigh the small percentage variation due to transition effects. 

![not much](https://user-images.githubusercontent.com/11737748/120161942-38fe9700-c23b-11eb-9c03-5300842bc31f.jpg)

Here I used extreme feedforward smoothing to make any added sharp steps visible, and moved the sticks as quickly as I could.

There was no sign of any spikes in this code or in master.  

![not much 2](https://user-images.githubusercontent.com/11737748/120163906-559bce80-c23d-11eb-881a-69b0941081d9.jpg)

Even though this code is perhaps a better way to do transition, its tough to see any problem from not doing it.

DusKing1 commented on src/main/flight/feedforward.c:
```diff
@@ -153,6 +154,10 @@ FAST_CODE_NOINLINE float feedforwardApply(int axis, bool newRcFrame, feedforward
         }
+
+        // apply feedforward transition
+        setpointDelta[axis] *= ffTransitionFactor > 0 ? MIN(1.f, getRcDeflectionAbs(axis) * ffTransitionFactor) : 1;
```
This line makes the pidControllerTest.testFeedForward in pid_unittest failed

ctzsnooze replied:
Thanks DusKing
I couldn't figure out why the change caused it to fail - not clever enough!
The test was expecting zero and it got an FF value, hence the failure.
But in actual testing the outcome was normal.
So I was not able to resolve this problem.

ctzsnooze replied:
I'm guessing that I need to include `feedforward.c` in the pid unit test?

DusKing1 replied:
maybe @mikeller ?

DusKing1 reviewed (commented):
I saw the unittest failed

haslinghuis commented on src/main/flight/feedforward.c:
```diff
@@ -153,6 +154,10 @@ FAST_CODE_NOINLINE float feedforwardApply(int axis, bool newRcFrame, feedforward
         }
+
+        // apply feedforward transition
+        setpointDelta[axis] *= ffTransitionFactor > 0 ? MIN(1.f, getRcDeflectionAbs(axis) * ffTransitionFactor) : 1;
```
For better readability I suggest `1.0f` instead of `1.f` (note the compiler does not care):
```suggestion
        setpointDelta[axis] *= ffTransitionFactor > 0 ? MIN(1.0f, getRcDeflectionAbs(axis) * ffTransitionFactor) : 1;
```

haslinghuis reviewed (changes requested):
@DusKing1 I don't experience errors with `make test` now

klutvott123 commented on src/main/flight/feedforward.c:
```diff
@@ -153,6 +154,10 @@ FAST_CODE_NOINLINE float feedforwardApply(int axis, bool newRcFrame, feedforward
         } else {
             setpointDelta[axis] = setpointDeltaImpl[axis];
         }
+
+        // apply feedforward transition
+        setpointDelta[axis] *= ffTransitionFactor > 0 ? MIN(1.0f, getRcDeflectionAbs(axis) * ffTransitionFactor) : 1;
+
```
This has to be be done here too for the unit tests to pass.
https://github.com/betaflight/betaflight/blob/0b0455d7db8a759da9070565e0f5651b1c2b73f2/src/test/unit/pid_unittest.cc#L100-L105

Something like this
```   
    float feedforwardApply(int axis, bool newRcFrame, feedforwardAveraging_t feedforwardAveraging)
    {
        UNUSED(newRcFrame);
        UNUSED(feedforwardAveraging);
        const float ffTransitionFactor = pidGetFfTransition();
        float setpointDelta = simulatedSetpointRate[axis] - simulatedPrevSetpointRate[axis];
        setpointDelta *= ffTransitionFactor > 0 ? MIN(1.0f, getRcDeflectionAbs(axis) * ffTransitionFactor) : 1;
        return setpointDelta;
    }
```

ctzsnooze replied:
Thanks @klutvott123 have merged those unit test fixes.

ctzsnooze commented:
Thanks to @klutvott123 the unit test issue is fixed.  

Now with this PR we only perform the transition mathematics 'on demand', ie when a new RC packet arrives, rather than every PID loop.

Should be ready to merge anytime.

haslinghuis reviewed (approved):
Unit test passed.

ctzsnooze commented:
I've rebased to the 4th August master and fixed the naming changes.
Also re-organised the feedforward related parameters into groups and, where appropriate, put them all inside `#ifdef USE_FEEDFORWARD`tags.

haslinghuis commented:
This needs a rebase

mikeller commented on src/main/flight/feedforward.c:
```diff
@@ -68,122 +65,134 @@ FAST_CODE_NOINLINE float feedforwardApply(int axis, bool newRcFrame, feedforward
-            // debug 2 is interpolated setpoint, above
-            DEBUG_SET(DEBUG_FF_INTERPOLATED, 3, lrintf(rcCommandDelta * 100.0f)); // rcCommand packet difference
+            DEBUG_SET(DEBUG_FEEDFORWARD, 1, lrintf(feedforward * 100.0f)); // delta after interpolating duplicates and smoothing
+            DEBUG_SET(DEBUG_FEEDFORWARD, 2, lrintf(boostAmount * 100.0f)); // boost amount after jitter reduction and smoothing
```
Probably better to keep this value in the same slot (1).

mikeller commented on src/main/flight/feedforward.c:
```diff
@@ -68,122 +65,134 @@ FAST_CODE_NOINLINE float feedforwardApply(int axis, bool newRcFrame, feedforward
 
         if (axis == FD_ROLL) {
-            DEBUG_SET(DEBUG_FF_INTERPOLATED, 2, lrintf(setpoint)); // setpoint after interpolations
+            DEBUG_SET(DEBUG_FEEDFORWARD, 0, lrintf(setpoint)); // setpoint after interpolations
```
Probably better to keep this value in the same slot (2).

mikeller commented on src/main/flight/pid.h:
```diff
@@ -206,10 +204,13 @@ typedef struct pidProfile_s {
-    uint8_t feedforward_max_rate_limit;     // Maximum setpoint rate percentage for feedforward
     uint8_t feedforward_smooth_factor;      // Amount of lowpass type smoothing for feedforward steps
     uint8_t feedforward_jitter_factor;      // Number of RC steps below which to attenuate feedforward
+    uint8_t feedforward_boost;              // amount of setpoint acceleration to add to feedforward, 10 means 100% added
```
There is no reason to move this around inside the parameter group - this is not how parameter groups work.

ctzsnooze replied:
If it does no harm, my intention was to get all the feedforward elements into the same order within the code, for ease of maintenance.

mikeller commented on src/main/flight/pid.h:
```diff
@@ -206,10 +204,13 @@ typedef struct pidProfile_s {
     uint8_t dyn_idle_d_gain;                // D gain for corrections around rapid changes in rpm
     uint8_t dyn_idle_max_increase;          // limit on maximum possible increase in motor idle drive during active control
 
+    uint8_t feedforwardTransition;          // Feedforward attenuation around centre sticks
```
This should go at the end of the parameter group. Also, since this is an array type parameter group, the version number will have to be incremented in https://github.com/betaflight/betaflight/pull/10777/files#diff-afd1226d5a8ae512e5fc53a545d9b7259b4f001bc94eae9c38cacd8762fb7f54R124.

ctzsnooze replied:
I am probably wrong, but for readability and maintenance, I thought it might help to keep all related parameters together.

mikeller commented on src/main/flight/pid.h:
```diff
@@ -206,10 +204,13 @@ typedef struct pidProfile_s {
     uint8_t feedforward_smooth_factor;      // Amount of lowpass type smoothing for feedforward steps
     uint8_t feedforward_jitter_factor;      // Number of RC steps below which to attenuate feedforward
+    uint8_t feedforward_boost;              // amount of setpoint acceleration to add to feedforward, 10 means 100% added
+    uint8_t feedforward_max_rate_limit;     // Maximum setpoint rate percentage for feedforward
```
There is no reason to move this around inside the parameter group - this is not how parameter groups work.

ctzsnooze replied:
If it does no harm, my intention was to get all the feedforward elements into the same order within the code, for ease of maintenance.

mikeller reviewed (changes requested):
Some things that have to be fixed.

ctzsnooze commented on src/main/cli/settings.c:
```diff
@@ -1150,12 +1149,13 @@ const clivalue_t valueTable[] = {
-    { "feedforward_smooth_factor",    VAR_UINT8 | PROFILE_VALUE, .config.minmaxUnsigned = {0, 75}, PG_PID_PROFILE, offsetof(pidProfile_t, feedforward_smooth_factor) },
+    { "feedforward_smoothing",        VAR_UINT8 | PROFILE_VALUE, .config.minmaxUnsigned = {0, 75}, PG_PID_PROFILE, offsetof(pidProfile_t, feedforward_smooth_factor) },
     { "feedforward_jitter_reduction", VAR_UINT8 | PROFILE_VALUE, .config.minmaxUnsigned = {0, 20}, PG_PID_PROFILE, offsetof(pidProfile_t, feedforward_jitter_factor) },
+    { "feedforward_boost",            VAR_UINT8 | PROFILE_VALUE,  .config.minmaxUnsigned = { 0, 50 }, PG_PID_PROFILE, offsetof(pidProfile_t, feedforward_boost) },
```
feedforward boost can't work without feedforward.  I'll put it within #ifdefs

ctzsnooze commented on src/main/cli/settings.c:
```diff
@@ -1150,12 +1149,13 @@ const clivalue_t valueTable[] = {
 #endif
 
 #ifdef USE_FEEDFORWARD
+    { "feedforward_transition",       VAR_UINT8  | PROFILE_VALUE, .config.minmaxUnsigned = { 0, 100 }, PG_PID_PROFILE, offsetof(pidProfile_t, feedforwardTransition) },
```
feedforward transition can't work without feedforward. 
I'll put it within #ifdefs now
